smart_card_connector_app/src/application_unittest.cc

```diff
@@ -1175,4 +1208,58 @@ TEST_F(SmartCardConnectorApplicationSingleClientTest, DisconnectAfterRemoving) {
   // fixture.
 }
 
+// `SCardTransmit()` calls from JS should be able to send a request APDU to the
+// card and receive a response. We use a fake PIV card in this test.
+TEST_F(SmartCardConnectorApplicationSingleClientTest, TransmitPivCommands) {
+  // Arrange: set up a reader and a card with a PIV profile.
+  TestingSmartCardSimulation::Device device;
+  device.id = 123;
+  device.type = TestingSmartCardSimulation::DeviceType::kGemaltoPcTwinReader;
+  device.card_type = TestingSmartCardSimulation::CardType::kCosmoId70;
+  device.card_profile =
+      TestingSmartCardSimulation::CardProfile::kCharismathicsPiv;
+  SetUsbDevices({device});
+  StartApplication();
+  SetUpJsClient();
+  SetUpSCardContext();
+  // Connect to the card.
+  SCARDHANDLE scard_handle = 0;
+  DWORD active_protocol = 0;
+  EXPECT_EQ(SimulateConnectCallFromJsClient(
+                kFakeHandlerId, scard_context(), kGemaltoPcTwinReaderPcscName0,
+                SCARD_SHARE_SHARED,
+                /*preferred_protocols=*/SCARD_PROTOCOL_T1, scard_handle,
+                active_protocol),
+            SCARD_S_SUCCESS);
+
+  // Act:
+  {
+    // Send the SELECT command (the format is per NIST 800-73-4).
+    const std::vector<uint8_t> kSelectCommand = {0x00, 0xA4, 0x04, 0x00, 0x09,
+                                                 0xA0, 0x00, 0x00, 0x03, 0x08,
+                                                 0x00, 0x00, 0x10, 0x00, 0x00};
+    std::vector<uint8_t> response;
+    DWORD response_protocol = 0;
+    EXPECT_EQ(
+        SimulateTransmitCallFromJsClient(
+            kFakeHandlerId, scard_handle, SCARD_PROTOCOL_T1, kSelectCommand,
+            /*receive_protocol=*/{}, response_protocol, response),
+        SCARD_S_SUCCESS);
+    EXPECT_EQ(response_protocol, static_cast<DWORD>(SCARD_PROTOCOL_T1));
+    // The expected result should contain the application identifier followed by
+    // 0x90 0x00 (denoting a successful operation).
+    std::vector<uint8_t> expected_response =
```

Comment:
In combination with the simulation which also calls `TestingSmartCardSimulation::GetCardProfileApplicationIdentifier` this feels a bit like doing `EXPECT_EQ(f(), f())`, but I don't have a idea how to test this in a better way either

Reply:
Yeah, I'm on the fence as well. At least it checks that the response correctly got through all the layers of the driver (which contains a lot of low-level handling).

I remember in some of the previous PRs there was a similar case where I initially hardcoded the same value in two places and we decided to share the constant. Maybe here in the current PR a duplicated constant would be more warranted.